File: B10 String/Leetcode 28 implement StrSTr[]/Leetcode.java
/*
Time Complexity:
Space Complexity:
this may cause time limited concern. better to ask how long the string is before you use this method
if the string is huge long. it is not a good way to solve

79 / 80 test cases passed.

*/ 


class Solution {
    public int strStr(String haystack, String needle) {
        if (needle == null || needle.length() == 0) return 0; 
        if (haystack == null || haystack.length() < needle.length()) return -1; 
        
        for(int i = 0 ; i < haystack.length(); i++){
            if(haystack.charAt(i) == needle.charAt(0)){
                int len = 0; 
                while(i+ len < haystack.length() && len < needle.length() && haystack.charAt(i+ len) == needle.charAt(len)){
                    len ++;   
                }
                if (len == needle.length()) return i; 
            }
            
        }
        return -1; 
        
    }
}

File: A01 Array/LC123 Best time to buy and sell stocks II/Leetcode.py
###########
Time Complexity: O(n)
Space Complexity:O(1)


#############
# A.1: Clearest Logic
class Solution:
    def maxProfit(self, prices):
        best = 0
        for i in range(1,len(prices)):
            if prices[i] > prices[i-1]:
                best += prices[i] - prices[i-1]
        return best